cpp/dolfinx/refinement/interval.h

```diff
@@ -0,0 +1,225 @@
+  MPI_Comm neighbor_comm;
+  MPI_Dist_graph_create_adjacent(
+      mesh.comm(), ranks.size(), ranks.data(), MPI_UNWEIGHTED, ranks.size(),
+      ranks.data(), MPI_UNWEIGHTED, MPI_INFO_NULL, false, &neighbor_comm);
+
+  // Communicate ghost cells that might have been marked. This is not necessary
+  // for a uniform refinement.
+  if (cells.has_value())
+    update_logical_edgefunction(neighbor_comm, marked_for_update,
+                                refinement_marker, *map_c);
+
+  // Construct the new vertices
+  const auto [new_vertex_map, new_vertex_coords, xshape]
+      = create_new_vertices(neighbor_comm, cell_ranks, mesh, refinement_marker);
+  MPI_Comm_free(&neighbor_comm);
+
+  auto c_to_v = mesh.topology()->connectivity(1, 0);
+  assert(c_to_v);
+
+  // get the count of cells to refine, note: we only consider non-ghost cells
+  std::int32_t number_of_refined_cells
+      = std::count(refinement_marker.begin(),
+                   std::next(refinement_marker.begin(),
+                             mesh.topology()->index_map(1)->size_local()),
+                   true);
+
+  // Produce local global indices, by padding out the previous index map
+  std::vector<std::int64_t> global_indices
+      = adjust_indices(*mesh.topology()->index_map(0), number_of_refined_cells);
+
+  // Build the topology on the new vertices
+  const auto refined_cell_count = mesh.topology()->index_map(1)->size_local()
+                                  + mesh.topology()->index_map(1)->num_ghosts()
+                                  + number_of_refined_cells;
+
+  std::vector<std::int64_t> cell_topology;
+  cell_topology.reserve(refined_cell_count * 2);
+
+  std::vector<std::int32_t> parent_cell;
+  parent_cell.reserve(refined_cell_count);
+
+  for (std::int32_t cell = 0; cell < map_c->size_local(); ++cell)
+  {
+    const auto& vertices = c_to_v->links(cell);
+    assert(vertices.size() == 2);
+
+    // we consider a (previous) cell, i.e. an edge of (global) vertices
+    // a ----------- b
+    const std::int64_t a = global_indices[vertices[0]];
+    const std::int64_t b = global_indices[vertices[1]];
+
+    if (refinement_marker[cell])
+    {
+      // find (global) index of new midpoint vertex:
+      // a --- c --- b
+      auto it = new_vertex_map.find(cell);
+      assert(it != new_vertex_map.end());
+      const std::int64_t c = it->second;
+
+      // add new cells/edges to refined topology
+      cell_topology.insert(cell_topology.end(), {a, c, c, b});
+      parent_cell.insert(parent_cell.end(), {cell, cell});
+    }
+    else
+    {
+      // copy the previous cell
+      cell_topology.insert(cell_topology.end(), {a, b});
+      parent_cell.push_back(cell);
+    }
+  }
+
+  assert(cell_topology.size() == refined_cell_count * 2);
+  assert(parent_cell.size() == refined_cell_count);
+
+  std::vector<std::int32_t> offsets(refined_cell_count + 1);
+  std::ranges::generate(offsets, [i = 0]() mutable { return 2 * i++; });
```

Comment:
Looks complex. Wouldn't iota work ok?

Reply:
Unfortunately `iota` only allows for step size one.

Reply:
Ah, it's always great to see the stuff that C++ has had to wait until C++23 to have...